### PR TITLE
Store entry assembly path for easy access for diagnostics

### LINKED ISSUE #94474: Make it easy to determine entry assembly path when debugging CoreCLR
When debugging for example crash dumps of .NET applications, it is often helpful to be able to quickly identify the entry assembly for the process/dump. Currently this information is easily accessible from managed code via `Assembly.GetEntryAssembly().Location` but it's very difficult to get to when in the debugger without ability to invoke methods or run managed code in general. For example, if the process was started via `dotnet app.dll`, it's difficult to determine the `app.dll` from a dump.

The proposal is to add a new static global variable into coreclr module, which will contain the path of the entry assembly once that information is known.

In detail:
* Add static global variable with the full path to the entry assembly
* The variable would be a `WCHAR *` type - UTF16 encoded string stored on the heap
* Modify `EnumMemRegions` such that string on the heap is stored in a dump even if the dump will not contain the full heap
* The variable is only initialized if the runtime "executes an assembly", so not if the runtime was started via native hosting APIs and is invoked using `get_delegate` or similar mechanisms.
* The variable is left `NULL` until it can be initialized, or for the lifetime of the runtime if there's no entry assembly defined (native hosting scenarios).
* The variable doesn't change its value once it's set to some path

To be figured out:
* The name of the variable
* Interaction with startup hooks - when exactly we would set the value (before or after startup hooks)

### PR DESCRIPTION
**Pull request by elinor-fung**
- Add `g_EntryAssemblyPath` global variable holding the full path to the entry assembly
  - Set right before loading the entry assembly (so also before startup hooks are run) - NULL if there is no entry assembly
- Ensure value is included dumps
  - For triage dumps, the dumped value is updated to only be the assembly file name instead of the full path

Example:
```
0:000> ||
.  0 User mini triage dump: C:\repos\helloworld\triage.dmp
0:000> x coreclr!g_EntryAssemblyPath
00007ffb`14f7dd58 coreclr!g_EntryAssemblyPath = 0x00000211`88016eb0 "helloworld.dll"

0:000> ||
.  0 User mini dump: C:\repos\helloworld\mini.dmp
0:000> x coreclr!g_EntryAssemblyPath
00007ffb`14f7dd58 coreclr!g_EntryAssemblyPath = 0x00000211`88016eb0 "C:\repos\helloworld\bin\Debug\net8.0\helloworld.dll"
```

Fixes https://github.com/dotnet/runtime/issues/94474

@leculver 
@dotnet/dotnet-diag

### CODE REVIEW

File: src/coreclr/debug/daccess/enummem.cpp [hunk 1]
```diff
@@ -224,6 +224,34 @@ HRESULT ClrDataAccess::EnumMemCLRStatic(IN CLRDataEnumMemoryFlags flags)
...
+                const WCHAR* name = u16_strrchr(path, DIRECTORY_SEPARATOR_CHAR_W);
+                if (name != NULL)
+                {
+                    name += 1;
+                    size_t len = u16_strlen(name) + 1;
+                    wcscpy_s(path, len, name);
+
+                    // Null out the rest of the buffer
+                    for (size_t i = len; i < pathLen; ++i)
+                        path[i] = '\0';
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/debug/daccess/enummem.cpp:248]:
```suggestion
                        path[i] = W('\0');
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by vitek-karas — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
FYI @dotnet/dotnet-diag

**Review by davmason — APPROVED** [review]

**Comment by mikem8361** [comment]
How is this supposed to be used?  It looks like a DAC'ized global variable was added (g_EntryAssemblyPath) but that is only accessible by the DAC and there is no DAC API exposed to query it.

**Comment by elinor-fung** [comment]
It is simply accessed as a global variable rather than via a DAC API.

**Comment by mikem8361** [comment]
The problem is that to access the global the tooling needs the symbol file.  It is was an export, then tools like createdump or dotnet-dump, etc. could use it.  I know that !analyze/Watson was it's intended usage so requiring a native debugger and symbols.

**Pull request merged**